Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2722,6 +2728,148 @@ extension BrowserViewController: ContextMenuHelperDelegate {
     }
 }
 
+/**
+ A third party search engine Browser extension
+**/
+extension BrowserViewController {
+
+    /**
+    This allows us to find the part of the WKWebview that has access to
+    the inputAcessoryView
+    **/
+    func findWebViewFirstResponder(view:UIView) -> UIView? {
+        let subviews = view.subviews
+        if subviews.count == 0 {
+            return nil
+        }
+        for subview: UIView in subviews {
+            if subview.isFirstResponder() {
+                return subview
+            }
+            return findWebViewFirstResponder(subview)
+        }
+        return nil
+    }
+
+    func addCustomSearchButtonToWebView(webView:WKWebView) {
+        if self.customSearchEngineButton == nil {
+            self.customSearchEngineButton = self.searchEngineButton()
+        }
+
+        //check if the search engine has already been added.
+        let domain = webView.URL?.domainURL().host
+        let matches = self.profile.searchEngines.orderedEngines.filter({$0.shortName == domain})
+        if (!matches.isEmpty) {
+            self.customSearchEngineButton!.tintColor = UIColor.grayColor()
+            self.customSearchEngineButton?.userInteractionEnabled = false
+        } else {
+            self.customSearchEngineButton!.tintColor = UIConstants.SystemBlueColor
+            self.customSearchEngineButton?.userInteractionEnabled = true
+        }
+
+        guard let webContentView = findWebViewFirstResponder(webView),
+            let input = webContentView.performSelector(Selector("inputAccessoryView")),
+            let inputView = input.takeUnretainedValue() as? UIInputView,
```

Comment:
The object that is returned from `performSelector` on the webContentView is from objc land and as such I need to either `takeUnretainedValue` in order to access the object. Not quite sure if there is a way around this but this was the only way I was able to access the inputView.